src/core/friendly_errors/param_validator.js

```diff
@@ -68,45 +132,95 @@ function generateZodSchemasForFunc(func) {
     overloads = funcInfo.overloads;
   }
 
-  const createParamSchema = param => {
+  // Generate a schema for a single parameter that can be of multiple constants.
+  // Note that because we're ultimately interested in the value of the constant,
+  // mapping constants to their values via `constantsMap` is necessary.
+  //
+  // Also, z.num() can only be used for a fixed set of allowable string values.
+  // However, our constants sometimes have numeric or non-primitive values.
+  // Therefore, z.union() is used here instead.
+  const generateConstantSchema = constants => {
+    return z.union(constants.map(c => z.literal(constantsMap[c])));
+  }
+
+  // Generate a schema for a single parameter that can be of multiple
+  // non-constant types, i.e. `String|Number|Array`.
+  const generateUnionSchema = types => {
+    return z.union(types
+      .filter(t => {
+        if (!schemaMap[t]) {
+          console.log(`Warning: Zod schema not found for type '${t}'. Skip mapping`);
+          return false;
+        }
+        return true;
+      })
+      .map(t => schemaMap[t]));
+  }
+
+  // Generate a schema for a single parameter.
+  const generateParamSchema = param => {
     const optional = param.endsWith('?');
     param = param.replace(/\?$/, '');
 
+    let schema;
+
+    // The parameter is can be of multiple types / constants
     if (param.includes('|')) {
       const types = param.split('|');
 
-      /*
-       * Note that for parameter types that are constants, such as for
-       * `blendMode`, the parameters are always all caps, sometimes with
-       * underscores, separated by `|`
-       * (i.e. "BLEND|DARKEST|LIGHTEST|DIFFERENCE|MULTIPLY|EXCLUSION|SCREEN|
-       * REPLACE|OVERLAY|HARD_LIGHT|SOFT_LIGHT|DODGE|BURN|ADD|REMOVE|SUBTRACT").
-       * Use a regex check here to filter them out and distinguish them from
-       * parameters that allow multiple types.
-       */
-      return types.every(t => /^[A-Z_]+$/.test(t))
-        ? z.enum(types)
-        : z.union(types
-          .filter(t => {
-            if (!schemaMap[t]) {
-              console.warn(`Warning: Zod schema not found for type '${t}'. Skip mapping`);
-              return false;
-            }
-            return true;
-          })
-          .map(t => schemaMap[t]));
+      // Note that for parameter types that are constants, such as for
+      // `blendMode`, the parameters are always all caps, sometimes with
+      // underscores, separated by `|` (i.e. "BLEND|DARKEST|LIGHTEST").
+      // Use a regex check here to filter them out and distinguish them from
+      // parameters that allow multiple types.
+      schema = types.every(t => /^[A-Z_]+$/.test(t))
+        ? generateConstantSchema(types)
+        : generateUnionSchema(types);
```

Comment:
Does this case also handle constants if they exist? e.g. something like `Number|AUTO` maybe? If not, maybe we can use `generateUnionSchema` for everything and update its `map` to map to a `z.literal` if a type is a constant.

Reply:
nice, looks good!